File: tests/test_static_files.py
#!/usr/bin/env python3
"""Integration tests for layer 5: serving files out of the document root
with sendfile(), and refusing to serve anything outside it.

The path-sanitization rules themselves are pinned down in
tests/test_static_files.c, which can enumerate hostile paths far more
cheaply. What only a running server can show is the rest: that the
headers describe the bytes that actually arrive, that a body too large
for the socket buffer still comes out whole (the EAGAIN / EPOLLOUT
backpressure path through sendfile), and that a client vanishing
mid-download doesn't take the process with it.

The server is pointed at a scratch document root rather than the
repository's public/, so the suite can put things in it that don't
belong in git: a multi-megabyte file, and a symlink escaping the root.
"""

import os
import shutil
import socket
import struct
import tempfile
import time

from http_test_utils import (ResponseReader, body_of, check, connect, header_value, parse_status, report,
                             running_server)

PUBLIC_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "public")

# Comfortably past any plausible socket send buffer, so the server is
# guaranteed to run out of room mid-file and have to resume on EPOLLOUT.
BIG_FILE_SIZE = 4 * 1024 * 1024
# Squeezing the client's receive window makes that certain rather than
# merely likely, and keeps the server parked in the resume path long
# enough for it to matter.
SMALL_RECV_BUFFER = 4096

root = None  # scratch document root, filled in by main()


def read_public_file(relative_path):
    with open(os.path.join(root, relative_path), "rb") as f:
        return f.read()


def get(path, extra_headers=b"", method=b"GET", timeout=2.0, recv_buffer_size=None):
    """One request on its own connection, closed on the way out."""
    with connect(recv_buffer_size) as s:
        reader = ResponseReader(s)
        s.sendall(method + b" " + path + b" HTTP/1.1\r\nHost: x\r\n" + extra_headers + b"\r\n")
        return reader.read_one(timeout=timeout)


def test_html_file_is_served_with_matching_headers_and_body():
    print("test_html_file_is_served_with_matching_headers_and_body")
    expected = read_public_file("index.html")
    r = get(b"/index.html")
    check(parse_status(r) == 200, "/index.html is 200")
    check(header_value(r, "Content-Type") == "text/html; charset=utf-8", "Content-Type comes from the .html extension")
    check(header_value(r, "Content-Length") == str(len(expected)), "Content-Length is the file size from fstat()")
    check(body_of(r) == expected, "body is the file byte for byte")


def test_text_file_is_served():
    print("test_text_file_is_served")
    expected = read_public_file("hello.txt")
    r = get(b"/hello.txt")
    check(parse_status(r) == 200, "/hello.txt is 200")
    check(header_value(r, "Content-Type") == "text/plain; charset=utf-8", "Content-Type comes from the .txt extension")
    check(body_of(r) == expected, "body is the file byte for byte")


def test_nested_path_is_served():
    print("test_nested_path_is_served")
    expected = read_public_file("assets/style.css")
    r = get(b"/assets/style.css")
    check(parse_status(r) == 200, "a file in a subdirectory is served")
    check(header_value(r, "Content-Type") == "text/css; charset=utf-8", "Content-Type comes from the .css extension")
    check(body_of(r) == expected, "body is the file byte for byte")


def test_unknown_extension_is_octet_stream():
    print("test_unknown_extension_is_octet_stream")
    r = get(b"/blob.unknown")
    check(parse_status(r) == 200, "a file with an unmapped extension is still served")
    check(header_value(r, "Content-Type") == "application/octet-stream",
          "unmapped extensions default to application/octet-stream instead of being sniffed")


def test_query_string_is_ignored_when_resolving():
    print("test_query_string_is_ignored_when_resolving")
    r = get(b"/hello.txt?v=2")
    check(parse_status(r) == 200, "a query string doesn't stop the file from being found")
    check(body_of(r) == read_public_file("hello.txt"), "and doesn't change which file it is")


def test_builtin_routes_still_win_over_the_filesystem():
    print("test_builtin_routes_still_win_over_the_filesystem")
    r = get(b"/")
    check(parse_status(r) == 200, "/ is still the built-in route")
    check(b"Hello from my C server!" in r, "/ is not resolved to public/index.html")
    r = get(b"/hello")
    check(b"Hello, hello!" in r, "/hello is still the built-in route")


def test_missing_file_is_404():
    print("test_missing_file_is_404")
    check(parse_status(get(b"/nope.txt")) == 404, "a file that isn't there is 404")
    check(parse_status(get(b"/assets/nope.css")) == 404, "same in a subdirectory")


def test_directory_is_404():
    print("test_directory_is_404")
    check(parse_status(get(b"/assets")) == 404, "a directory is 404, there are no listings")
    check(parse_status(get(b"/assets/")) == 404, "and a trailing slash doesn't change that")


def test_path_traversal_is_400():
    print("test_path_traversal_is_400")
    # 400 rather than 404 because a ".." segment makes the request itself
    # defective: no legitimate client sends one, and the server never
    # looked at the filesystem, so "not found" would be asserting
    # something it never checked. See static_file_open() in
    # src/static_files.c.
    check(parse_status(get(b"/../Makefile")) == 400, "a literal ../ traversal is 400")
    check(parse_status(get(b"/../../etc/passwd")) == 400, "climbing several levels is 400")
    check(parse_status(get(b"/assets/../hello.txt")) == 400,
          "even a ../ that would have landed back inside the root is rejected")


def test_encoded_path_traversal_is_400():
    print("test_encoded_path_traversal_is_400")
    # The bypass a resolver that inspects the raw request-target falls
    # for: the dots only look like dots after percent-decoding, which is
    # exactly what open() would have seen.
    check(parse_status(get(b"/%2e%2e/Makefile")) == 400, "%2e%2e is decoded before the .. check, so it's 400")
    check(parse_status(get(b"/..%2fMakefile")) == 400, "an encoded separator doesn't hide the traversal either")
    check(parse_status(get(b"/%2e%2e%2fsrc%2fmain.c")) == 400, "fully encoded traversal is 400")


def test_traversal_response_does_not_leak_the_file():
    print("test_traversal_response_does_not_leak_the_file")
    r = get(b"/../Makefile")
    check(b"gcc" not in r and b"CFLAGS" not in r, "the rejected response contains none of the target file")


def test_malformed_percent_encoding_is_400():
    print("test_malformed_percent_encoding_is_400")
    check(parse_status(get(b"/hello%2")) == 400, "a truncated escape is 400")
    check(parse_status(get(b"/hello%zz.txt")) == 400, "a non-hex escape is 400")
    check(parse_status(get(b"/hello.txt%00.png")) == 400, "an encoded NUL is 400, never truncated at")


def test_symlink_escaping_the_root_is_404():
    print("test_symlink_escaping_the_root_is_404")
    # Nothing in this request-target is suspicious, so the ".." rule
    # can't help; the containment check on the opened fd is what stops
    # it. 404 and not 400 precisely because the request was well formed:
    # answering anything else would confirm the target exists.
    r = get(b"/escape.txt")
    check(parse_status(r) == 404, "a symlink pointing outside the document root is 404")
    check(b"root:" not in r and b"127.0.0.1" not in r, "and none of the target file comes back")


def test_symlink_inside_the_root_is_served():
    print("test_symlink_inside_the_root_is_served")
    r = get(b"/link.txt")
    check(parse_status(r) == 200, "a symlink that stays inside the root is served normally")
    check(body_of(r) == read_public_file("hello.txt"), "and serves the file it points at")


def test_non_get_method_on_a_file_is_405():
    print("test_non_get_method_on_a_file_is_405")
    r = get(b"/index.html", method=b"POST")
    check(parse_status(r) == 405, "POST to an existing file is 405")
    check(header_value(r, "Allow") == "GET", "405 says which method would have worked")
    # 404 outranks 405: claiming "method not allowed" would confirm that
    # a resource exists there.
    check(parse_status(get(b"/nope.txt", method=b"POST")) == 404, "POST to a missing file is 404, not 405")


def test_traversal_does_not_break_the_connection():
    print("test_traversal_does_not_break_the_connection")
    # A rejected path is not a framing error: the request parsed
    # perfectly, so the connection stays usable, unlike the 400 for a
    # malformed request (see tests/test_keepalive.py).
    with connect() as s:
        reader = ResponseReader(s)
        s.sendall(b"GET /../Makefile HTTP/1.1\r\nHost: x\r\n\r\n")
        check(parse_status(reader.read_one()) == 400, "traversal attempt is answered 400")
        s.sendall(b"GET /hello.txt HTTP/1.1\r\nHost: x\r\n\r\n")
        check(parse_status(reader.read_one()) == 200, "the same connection still serves the next request")


def test_large_file_survives_backpressure():
    print("test_large_file_survives_backpressure")
    expected = read_public_file("big.bin")
    with connect(SMALL_RECV_BUFFER) as s:
        reader = ResponseReader(s)
        s.sendall(b"GET /big.bin HTTP/1.1\r\nHost: x\r\n\r\n")
        # Let the server fill the socket and hit EAGAIN in sendfile()
        # before a single byte is drained, so the rest of the transfer
        # has to come from the EPOLLOUT resume path.
        time.sleep(0.3)
        r = reader.read_one(timeout=20.0)
    check(parse_status(r) == 200, f"a {BIG_FILE_SIZE // (1024 * 1024)}MiB file is served")
    check(header_value(r, "Content-Length") == str(BIG_FILE_SIZE), "Content-Length matches the file size")
    check(body_of(r) == expected, "every byte arrives, in order, across the sendfile() resumes")


def test_keep_alive_still_works_after_a_large_file():
    print("test_keep_alive_still_works_after_a_large_file")
    # The interesting part is the bookkeeping: the file fd has to be
    # released and the parser reset before the next request, even though
    # this response finished in the EPOLLOUT handler rather than inline.
    with connect(SMALL_RECV_BUFFER) as s:
        reader = ResponseReader(s)
        s.sendall(b"GET /big.bin HTTP/1.1\r\nHost: x\r\n\r\n")
        time.sleep(0.3)
        r1 = reader.read_one(timeout=20.0)
        check(len(body_of(r1)) == BIG_FILE_SIZE, "large file fully received")

        s.sendall(b"GET /hello.txt HTTP/1.1\r\nHost: x\r\n\r\n")
        r2 = reader.read_one()
        check(parse_status(r2) == 200, "the same connection serves a second request afterwards")
        check(body_of(r2) == read_public_file("hello.txt"), "and serves the right file")


def test_pipelined_static_requests():
    print("test_pipelined_static_requests")
    with connect() as s:
        reader = ResponseReader(s)
        s.sendall(b"GET /hello.txt HTTP/1.1\r\n\r\nGET /assets/style.css HTTP/1.1\r\n\r\n")
        r1 = reader.read_one()
        r2 = reader.read_one()
        check(body_of(r1) == read_public_file("hello.txt"), "first pipelined file response is correct")
        check(body_of(r2) == read_public_file("assets/style.css"), "second pipelined file response is correct")


def test_client_disconnecting_mid_download_does_not_kill_the_server():
    print("test_client_disconnecting_mid_download_does_not_kill_the_server")
    # Writing to a socket the peer has reset raises SIGPIPE, whose
    # default action would end the process; sendfile() has no
    # MSG_NOSIGNAL, so main() ignores the signal outright. Without that,
    # this test kills the server and every check after it fails.
    s = connect(SMALL_RECV_BUFFER)
    s.sendall(b"GET /big.bin HTTP/1.1\r\nHost: x\r\n\r\n")
    s.recv(1024)  # let the transfer get going
    # SO_LINGER with a zero timeout makes close() send an RST instead of
    # a FIN, so the server's next sendfile() fails hard (EPIPE /
    # ECONNRESET) rather than draining politely into a half-closed
    # socket.
    s.setsockopt(socket.SOL_SOCKET, socket.SO_LINGER, struct.pack("ii", 1, 0))
    s.close()

    time.sleep(0.3)
    r = get(b"/hello.txt")
    check(parse_status(r) == 200, "the server is still serving after a client vanished mid-download")


def build_scratch_root(root_dir):
    """public/ plus the things that can't live in the repository."""
    for name in os.listdir(PUBLIC_DIR):
        source = os.path.join(PUBLIC_DIR, name)
        destination = os.path.join(root_dir, name)
        if os.path.isdir(source):
            shutil.copytree(source, destination)
        else:
            shutil.copy(source, destination)

    with open(os.path.join(root_dir, "blob.unknown"), "wb") as f:
        f.write(b"no extension mapping for this one\n")

    # Deterministic, non-repeating content: a file of identical bytes
    # would still look intact if the server duplicated or dropped a
    # chunk while resuming a partial sendfile().
    with open(os.path.join(root_dir, "big.bin"), "wb") as f:
        f.write(bytes((i * 7 + i // 251) % 256 for i in range(BIG_FILE_SIZE)))

    os.symlink("hello.txt", os.path.join(root_dir, "link.txt"))
    # Absolute, and pointing at a file that exists on any Linux box.
    os.symlink("/etc/hosts", os.path.join(root_dir, "escape.txt"))


def main():
    global root
    with tempfile.TemporaryDirectory(prefix="http-server-static-") as root_dir:
        root = root_dir
        build_scratch_root(root_dir)

        with running_server({"HTTP_SERVER_PUBLIC_ROOT": root_dir}):
            test_html_file_is_served_with_matching_headers_and_body()
            test_text_file_is_served()
            test_nested_path_is_served()
            test_unknown_extension_is_octet_stream()
            test_query_string_is_ignored_when_resolving()
            test_builtin_routes_still_win_over_the_filesystem()
            test_missing_file_is_404()
            test_directory_is_404()
            test_path_traversal_is_400()
            test_encoded_path_traversal_is_400()
            test_traversal_response_does_not_leak_the_file()
            test_malformed_percent_encoding_is_400()
            test_symlink_escaping_the_root_is_404()
            test_symlink_inside_the_root_is_served()
            test_non_get_method_on_a_file_is_405()
            test_traversal_does_not_break_the_connection()
            test_large_file_survives_backpressure()
            test_keep_alive_still_works_after_a_large_file()
            test_pipelined_static_requests()
            test_client_disconnecting_mid_download_does_not_kill_the_server()

    report()


if __name__ == "__main__":
    main()

File: tests/test_static_files.c
/* Unit tests for the static-file resolver, mostly for the path
 * sanitization: traversal handling is the kind of thing that must be
 * pinned down by tests that can enumerate hostile inputs cheaply,
 * without a socket or a server process in the way.
 *
 * Two roots are used. The repository's own public/ checks the files the
 * server actually ships (so a renamed file or a lost extension mapping
 * shows up here), and a scratch directory built at runtime covers what
 * can't be committed to git: symlinks that escape the root, FIFOs, and
 * a sibling directory whose name shares a prefix with the root.
 *
 * Same house style as tests/test_http_parser.c: no framework, just
 * CHECK. */

#define _XOPEN_SOURCE 700

#include <limits.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

#include "static_files.h"

static int g_tests_run = 0;
static int g_tests_failed = 0;

#define CHECK(cond)                                                                                                  \
    do {                                                                                                             \
        g_tests_run++;                                                                                               \
        if (!(cond)) {                                                                                               \
            g_tests_failed++;                                                                                        \
            fprintf(stderr, "  FAIL %s:%d: %s\n", __FILE__, __LINE__, #cond);                                        \
        }                                                                                                            \
    } while (0)

#define RUN(test_fn)                                                                                                 \
    do {                                                                                                             \
        printf("%s\n", #test_fn);                                                                                    \
        test_fn();                                                                                                   \
    } while (0)

/* Resolves path and immediately closes whatever came back, for the many
 * tests that only care about the verdict. */
static static_file_result_t resolve(const char *path) {
    static_file_t file;
    static_file_result_t result = static_file_open(path, strlen(path), &file);
    static_file_close(&file);
    return result;
}

/* --- tests against the repository's public/ ------------------------- */

static void test_existing_files_are_served_with_a_content_type(void) {
    static_file_t file;

    CHECK(static_file_open("/hello.txt", strlen("/hello.txt"), &file) == STATIC_FILE_OK);
    CHECK(file.fd >= 0);
    CHECK(file.size > 0);
    CHECK(file.content_type != NULL && strcmp(file.content_type, "text/plain; charset=utf-8") == 0);
    static_file_close(&file);
    CHECK(file.fd == -1); /* close must clear the fd, callers use it as the "is there a body" flag */

    CHECK(static_file_open("/index.html", strlen("/index.html"), &file) == STATIC_FILE_OK);
    CHECK(strcmp(file.content_type, "text/html; charset=utf-8") == 0);
    static_file_close(&file);

    /* Nested path, and a type that is neither of the two obvious ones. */
    CHECK(static_file_open("/assets/style.css", strlen("/assets/style.css"), &file) == STATIC_FILE_OK);
    CHECK(strcmp(file.content_type, "text/css; charset=utf-8") == 0);
    static_file_close(&file);
}

static void test_missing_file_is_not_found(void) {
    CHECK(resolve("/does-not-exist.txt") == STATIC_FILE_NOT_FOUND);
    CHECK(resolve("/assets/nope.css") == STATIC_FILE_NOT_FOUND);
    /* A missing intermediate directory makes open() fail with ENOTDIR
     * rather than ENOENT, same answer either way. */
    CHECK(resolve("/hello.txt/deeper") == STATIC_FILE_NOT_FOUND);
}

static void test_directories_are_not_served(void) {
    CHECK(resolve("/assets") == STATIC_FILE_NOT_FOUND);
    CHECK(resolve("/assets/") == STATIC_FILE_NOT_FOUND);
    CHECK(resolve("/") == STATIC_FILE_NOT_FOUND);
    /* A trailing slash names a directory, so it must not quietly serve
     * the file of the same name: one resource, one URL. */
    CHECK(resolve("/hello.txt/") == STATIC_FILE_NOT_FOUND);
}

static void test_query_string_and_fragment_are_ignored(void) {
    CHECK(resolve("/hello.txt?v=2") == STATIC_FILE_OK);
    CHECK(resolve("/hello.txt?") == STATIC_FILE_OK);
    CHECK(resolve("/hello.txt#top") == STATIC_FILE_OK);
}

static void test_redundant_path_syntax_is_normalized(void) {
    CHECK(resolve("/./hello.txt") == STATIC_FILE_OK);
    CHECK(resolve("//hello.txt") == STATIC_FILE_OK);
    CHECK(resolve("/assets/./style.css") == STATIC_FILE_OK);
    CHECK(resolve("//assets//style.css") == STATIC_FILE_OK);
}

static void test_percent_encoding_is_decoded(void) {
    CHECK(resolve("/hello%2Etxt") == STATIC_FILE_OK);  /* '.' */
    CHECK(resolve("/assets%2fstyle.css") == STATIC_FILE_OK); /* an encoded separator is still a separator */
}

/* The core of layer 5's security story: none of these may reach a file. */
static void test_path_traversal_is_rejected(void) {
    CHECK(resolve("/../Makefile") == STATIC_FILE_INVALID_PATH);
    CHECK(resolve("/../../etc/passwd") == STATIC_FILE_INVALID_PATH);
    CHECK(resolve("/assets/../hello.txt") == STATIC_FILE_INVALID_PATH);
    CHECK(resolve("/..") == STATIC_FILE_INVALID_PATH);
    CHECK(resolve("/a/b/../../../etc/passwd") == STATIC_FILE_INVALID_PATH);
}

/* Decoding happens before the ".." check, so encoding the dots buys the
 * attacker nothing. A resolver that checked the raw target first would
 * let every one of these through. */
static void test_encoded_path_traversal_is_rejected(void) {
    CHECK(resolve("/%2e%2e/Makefile") == STATIC_FILE_INVALID_PATH);
    CHECK(resolve("/%2E%2E/Makefile") == STATIC_FILE_INVALID_PATH);
    CHECK(resolve("/..%2fMakefile") == STATIC_FILE_INVALID_PATH);
    CHECK(resolve("/%2e%2e%2fMakefile") == STATIC_FILE_INVALID_PATH);
    CHECK(resolve("/assets/%2e%2e/%2e%2e/Makefile") == STATIC_FILE_INVALID_PATH);
    /* Double encoding: "%252e" decodes to the literal text "%2e", which
     * is a filename, not a dot. One decoding pass only, so this is an
     * ordinary miss rather than a traversal. */
    CHECK(resolve("/%252e%252e/Makefile") == STATIC_FILE_NOT_FOUND);
}

static void test_malformed_percent_encoding_is_rejected(void) {
    CHECK(resolve("/hello%2") == STATIC_FILE_INVALID_PATH);  /* truncated escape */
    CHECK(resolve("/hello%") == STATIC_FILE_INVALID_PATH);
    CHECK(resolve("/hello%zz.txt") == STATIC_FILE_INVALID_PATH); /* not hex */
    CHECK(resolve("/hello%2g.txt") == STATIC_FILE_INVALID_PATH);
}

/* "%00" is the classic truncation trick: length-based checks see the
 * full name, open() stops at the NUL and gets a different file. */
static void test_encoded_nul_is_rejected(void) {
    CHECK(resolve("/hello.txt%00.png") == STATIC_FILE_INVALID_PATH);
    CHECK(resolve("/%00") == STATIC_FILE_INVALID_PATH);
}

static void test_non_absolute_path_is_rejected(void) {
    CHECK(resolve("hello.txt") == STATIC_FILE_INVALID_PATH);
    CHECK(resolve("") == STATIC_FILE_INVALID_PATH);
    /* Absolute-form targets ("GET http://host/x") aren't origin-form and
     * this server doesn't accept them as file paths either. */
    CHECK(resolve("http://example.com/hello.txt") == STATIC_FILE_INVALID_PATH);
}

static void test_overlong_path_is_a_miss_not_an_error(void) {
    char path[PATH_MAX * 2];
    memset(path, 'a', sizeof(path) - 1);
    path[0] = '/';
    path[sizeof(path) - 1] = '\0';
    /* Nothing hostile about it, it simply cannot name anything under a
     * root that leaves less room than this. */
    CHECK(resolve(path) == STATIC_FILE_NOT_FOUND);
}

/* --- tests against a scratch root built at runtime ------------------ */

/* The scratch root is always a mkdtemp() of this one template, and the
 * sibling is that same name plus a short suffix, so both have a length
 * known at compile time. Sizing them at PATH_MAX instead would be a
 * claim the code never makes good on, and it costs something real: the
 * compiler then has to assume a 4095-character root, which makes every
 * "%s/<name>" below look like it might not fit its PATH_MAX destination
 * (-Wformat-truncation, visible once optimization is on). Right-sizing
 * these is what lets it prove the concatenations fit. */
#define SCRATCH_TEMPLATE "/tmp/http-server-static-XXXXXX"
#define SCRATCH_SIBLING_SUFFIX "evil"
#define SCRATCH_ROOT_SIZE (sizeof(SCRATCH_TEMPLATE) + sizeof(SCRATCH_SIBLING_SUFFIX))

static char g_scratch[SCRATCH_ROOT_SIZE];
static char g_scratch_sibling[SCRATCH_ROOT_SIZE];

static void scratch_path(char *out, size_t out_size, const char *name) {
    snprintf(out, out_size, "%s/%s", g_scratch, name);
}

static bool write_file(const char *full_path, const char *contents) {
    FILE *f = fopen(full_path, "w");
    if (f == NULL) {
        perror("fopen");
        return false;
    }
    fputs(contents, f);
    return fclose(f) == 0;
}

/* Builds:
 *   <scratch>/plain.dat          regular file, unknown extension
 *   <scratch>/inside.txt         regular file
 *   <scratch>/link-inside        symlink -> inside.txt        (must be served)
 *   <scratch>/link-escape        symlink -> /etc/hostname     (must not be)
 *   <scratch>/link-sibling       symlink -> <scratch>evil/f   (prefix trap)
 *   <scratch>/fifo               FIFO, which sendfile() cannot read
 *   <scratch>evil/secret.txt     outside the root, shares its name prefix
 */
static bool setup_scratch_root(void) {
    char template[] = SCRATCH_TEMPLATE;
    if (mkdtemp(template) == NULL) {
        perror("mkdtemp");
        return false;
    }
    snprintf(g_scratch, sizeof(g_scratch), "%s", template);
    snprintf(g_scratch_sibling, sizeof(g_scratch_sibling), "%s" SCRATCH_SIBLING_SUFFIX, template);
    if (mkdir(g_scratch_sibling, 0700) < 0) {
        perror("mkdir(sibling)");
        return false;
    }

    char path[PATH_MAX];
    char target[PATH_MAX];

    scratch_path(path, sizeof(path), "plain.dat");
    if (!write_file(path, "binary-ish\n")) {
        return false;
    }
    scratch_path(path, sizeof(path), "inside.txt");
    if (!write_file(path, "inside the root\n")) {
        return false;
    }
    snprintf(path, sizeof(path), "%s/secret.txt", g_scratch_sibling);
    if (!write_file(path, "outside the root\n")) {
        return false;
    }

    scratch_path(path, sizeof(path), "link-inside");
    if (symlink("inside.txt", path) < 0) {
        perror("symlink(inside)");
        return false;
    }
    scratch_path(path, sizeof(path), "link-escape");
    if (symlink("/etc/hostname", path) < 0) {
        perror("symlink(escape)");
        return false;
    }
    scratch_path(path, sizeof(path), "link-sibling");
    snprintf(target, sizeof(target), "%s/secret.txt", g_scratch_sibling);
    if (symlink(target, path) < 0) {
        perror("symlink(sibling)");
        return false;
    }
    scratch_path(path, sizeof(path), "fifo");
    if (mkfifo(path, 0600) < 0) {
        perror("mkfifo");
        return false;
    }
    return true;
}

static void remove_scratch_root(void) {
    static const char *names[] = {"plain.dat", "inside.txt", "link-inside", "link-escape", "link-sibling", "fifo"};
    char path[PATH_MAX];
    for (size_t i = 0; i < sizeof(names) / sizeof(names[0]); i++) {
        scratch_path(path, sizeof(path), names[i]);
        unlink(path);
    }
    rmdir(g_scratch);

    snprintf(path, sizeof(path), "%s/secret.txt", g_scratch_sibling);
    unlink(path);
    rmdir(g_scratch_sibling);
}

static void test_unknown_extension_falls_back_to_octet_stream(void) {
    static_file_t file;
    CHECK(static_file_open("/plain.dat", strlen("/plain.dat"), &file) == STATIC_FILE_OK);
    CHECK(strcmp(file.content_type, "application/octet-stream") == 0);
    static_file_close(&file);
}

/* Symlinks are fine as long as what they point at stays inside: the
 * containment check looks at the file that was actually opened, not at
 * the name the client typed. */
static void test_symlink_inside_the_root_is_served(void) {
    CHECK(resolve("/link-inside") == STATIC_FILE_OK);
}

/* The one escape the ".." rejection cannot catch, since nothing in the
 * request path is suspicious at all. */
static void test_symlink_escaping_the_root_is_not_found(void) {
    /* 404 rather than 400 here: the request was well formed, and
     * answering "not found" declines to confirm that /etc/hostname
     * exists. */
    CHECK(resolve("/link-escape") == STATIC_FILE_NOT_FOUND);
}

/* Catches a containment check written as a bare strncmp(): "/tmp/xevil"
 * starts with "/tmp/x" but is not under it. */
static void test_sibling_directory_sharing_the_root_prefix_is_not_found(void) {
    CHECK(resolve("/link-sibling") == STATIC_FILE_NOT_FOUND);
}

/* sendfile(2) needs an in_fd that supports mmap-style reads, which rules
 * out FIFOs; opening one would also have blocked the event loop if the
 * resolver hadn't passed O_NONBLOCK. */
static void test_fifo_is_not_served(void) {
    CHECK(resolve("/fifo") == STATIC_FILE_NOT_FOUND);
}

/* --- root configuration --------------------------------------------- */

static void test_init_rejects_a_root_that_is_not_a_directory(void) {
    CHECK(static_files_init("public/hello.txt") < 0);
    CHECK(static_files_init("public/definitely-not-here") < 0);
}

int main(void) {
    /* Relative to the working directory, so this must run from the
     * repository root, which is how the Makefile invokes it. */
    if (static_files_init("public") < 0) {
        fprintf(stderr, "could not init static root 'public' (run from the repository root)\n");
        return 1;
    }

    RUN(test_existing_files_are_served_with_a_content_type);
    RUN(test_missing_file_is_not_found);
    RUN(test_directories_are_not_served);
    RUN(test_query_string_and_fragment_are_ignored);
    RUN(test_redundant_path_syntax_is_normalized);
    RUN(test_percent_encoding_is_decoded);
    RUN(test_path_traversal_is_rejected);
    RUN(test_encoded_path_traversal_is_rejected);
    RUN(test_malformed_percent_encoding_is_rejected);
    RUN(test_encoded_nul_is_rejected);
    RUN(test_non_absolute_path_is_rejected);
    RUN(test_overlong_path_is_a_miss_not_an_error);
    RUN(test_init_rejects_a_root_that_is_not_a_directory);

    if (!setup_scratch_root()) {
        fprintf(stderr, "could not build the scratch root\n");
        remove_scratch_root();
        return 1;
    }
    if (static_files_init(g_scratch) < 0) {
        remove_scratch_root();
        return 1;
    }

    RUN(test_unknown_extension_falls_back_to_octet_stream);
    RUN(test_symlink_inside_the_root_is_served);
    RUN(test_symlink_escaping_the_root_is_not_found);
    RUN(test_sibling_directory_sharing_the_root_prefix_is_not_found);
    RUN(test_fifo_is_not_served);

    remove_scratch_root();

    printf("\n%d checks run, %d failed\n", g_tests_run, g_tests_failed);
    return g_tests_failed == 0 ? 0 : 1;
}
